File: implementations/python/password_reset_proof.py
"""End-to-end proof runner for Orin password-reset meaning derivation."""

from copy import deepcopy
import json
from pathlib import Path
from typing import Any

from conformance_runner import execute_case, expected_case_result, load_cases
from lowering import lower
from orin_model import SemanticModel


ROOT = Path(__file__).parents[2]
MODEL_FIXTURE = ROOT / "tests" / "conformance" / "password-reset.model.json"
CASES_FIXTURE = ROOT / "tests" / "conformance" / "password-reset.cases.json"
POLICIES_FIXTURE = ROOT / "tests" / "conformance" / "password-reset.policies.json"


def _resolve_rate_limit(model: SemanticModel) -> SemanticModel:
    resolved = deepcopy(model.document)
    for obj in resolved.get("objects", []):
        if obj.get("kind") != "uncertainty":
            continue
        if obj.get("id") == "account.password-reset/uncertainty/rate-limit":
            obj["status"] = "resolved"
    return SemanticModel(resolved)


def _behavioral_case_results(model: SemanticModel, cases: dict[str, Any]) -> list[dict[str, Any]]:
    results: list[dict[str, Any]] = []
    for case in cases["cases"]:
        if case["when"]["action"] == "compile":
            continue
        actual = execute_case(model, case)
        expected = expected_case_result(case)
        for key, value in expected.items():
            if actual[key] != value:
                raise AssertionError(f"case {case['id']} mismatch for '{key}': {actual[key]} != {value}")
        results.append({"id": case["id"], "actual": actual})
    return results


def run_password_reset_proof() -> dict[str, Any]:
    blocked_model = SemanticModel.from_json_file(MODEL_FIXTURE)
    cases = load_cases(CASES_FIXTURE)
    policies = load_cases(POLICIES_FIXTURE)

    blocked_case = next(case for case in cases["cases"] if case["when"]["action"] == "compile")
    blocked_compile = execute_case(blocked_model, blocked_case)["compilation"]

    resolved_model = _resolve_rate_limit(blocked_model)
    resolved_status = resolved_model.compilation_status()
    if resolved_status != "eligible":
        raise AssertionError(f"resolved model should be eligible, got {resolved_status}")

    variant_outputs: list[dict[str, Any]] = []
    for variant in policies["variants"]:
        variant_doc = deepcopy(resolved_model.document)
        variant_doc["implementationPolicies"] = variant["implementationPolicies"]
        variant_model = SemanticModel(variant_doc)
        lowered = lower(variant_model)
        behavioral_results = _behavioral_case_results(variant_model, cases)
        variant_outputs.append({"id": variant["id"], "derived": lowered["artifact"], "behavior": behavioral_results})

    if variant_outputs[0]["derived"] == variant_outputs[1]["derived"]:
        raise AssertionError("derived artifacts should differ across policy variants")
    if variant_outputs[0]["behavior"] != variant_outputs[1]["behavior"]:
        raise AssertionError("observable behavior should remain equal across policy variants")

    return {
        "blockedCompilation": blocked_compile,
        "resolvedCompilation": resolved_status,
        "canonicalMeaningStableAcrossVariants": SemanticModel({
            **resolved_model.document,
            "implementationPolicies": policies["variants"][0]["implementationPolicies"],
        }).canonical() == SemanticModel({
            **resolved_model.document,
            "implementationPolicies": policies["variants"][1]["implementationPolicies"],
        }).canonical(),
        "variants": [
            {"id": item["id"], "derivedArtifact": item["derived"]} for item in variant_outputs
        ],
        "behaviorCases": [item["id"] for item in variant_outputs[0]["behavior"]],
    }


def main() -> int:
    print(json.dumps(run_password_reset_proof(), indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: implementations/typescript/src/password_reset_proof.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

import { executeCase } from "./conformance_runner.ts";
import { lower } from "./lowering.ts";
import { SemanticModel } from "./orin_model.ts";

const ROOT = resolve(fileURLToPath(new URL("../../..", import.meta.url)));
const MODEL_FIXTURE = resolve(ROOT, "tests/conformance/password-reset.model.json");
const CASES_FIXTURE = resolve(ROOT, "tests/conformance/password-reset.cases.json");
const POLICIES_FIXTURE = resolve(ROOT, "tests/conformance/password-reset.policies.json");

function loadJson(path: string): Record<string, any> {
  return JSON.parse(readFileSync(path, "utf-8"));
}

function resolveRateLimit(model: SemanticModel): SemanticModel {
  const next = JSON.parse(JSON.stringify(model.document));
  for (const obj of next.objects || []) {
    if (obj.kind === "uncertainty" && obj.id === "account.password-reset/uncertainty/rate-limit") {
      obj.status = "resolved";
    }
  }
  return new SemanticModel(next);
}

function deepEqual(left: unknown, right: unknown): boolean {
  if (left === right) {
    return true;
  }
  if (typeof left !== "object" || left === null || typeof right !== "object" || right === null) {
    return false;
  }
  if (Array.isArray(left) || Array.isArray(right)) {
    if (!Array.isArray(left) || !Array.isArray(right) || left.length !== right.length) {
      return false;
    }
    return left.every((item, index) => deepEqual(item, right[index]));
  }
  const leftEntries = Object.entries(left as Record<string, unknown>);
  const rightEntries = Object.entries(right as Record<string, unknown>);
  if (leftEntries.length !== rightEntries.length) {
    return false;
  }
  const rightMap = new Map<string, unknown>(rightEntries);
  return leftEntries.every(([key, value]) => rightMap.has(key) && deepEqual(value, rightMap.get(key)));
}

function behaviorResults(model: SemanticModel, casesFixture: Record<string, any>): Array<Record<string, any>> {
  const results: Array<Record<string, any>> = [];
  for (const conformanceCase of casesFixture.cases || []) {
    if (conformanceCase?.when?.action === "compile") {
      continue;
    }
    const actual = executeCase(model, conformanceCase);
    for (const [key, expected] of Object.entries(conformanceCase.then || {})) {
      if (!deepEqual(actual[key], expected)) {
        throw new Error(`case ${conformanceCase.id} mismatch for '${key}'`);
      }
    }
    results.push({ id: conformanceCase.id });
  }
  return results;
}

export function runDerivationProof(): Record<string, any> {
  const blockedModel = new SemanticModel(loadJson(MODEL_FIXTURE));
  const casesFixture = loadJson(CASES_FIXTURE);
  const policiesFixture = loadJson(POLICIES_FIXTURE);
  const variants = policiesFixture.variants || [];
  if (variants.length < 2) {
    throw new Error("proof requires at least two policy variants");
  }

  const blockedCase = (casesFixture.cases || []).find(
    (item: Record<string, any>) => item?.when?.action === "compile" && item?.then?.compilation === "blocked",
  );
  if (!blockedCase) {
    throw new Error("cases fixture must include blocked compile case");
  }

  const blockedCompilation = executeCase(blockedModel, blockedCase).compilation;
  const resolvedModel = resolveRateLimit(blockedModel);
  const resolvedCompilation = resolvedModel.compilationStatus();
  if (resolvedCompilation !== "eligible") {
    throw new Error(`resolved model should be eligible, got ${resolvedCompilation}`);
  }

  const variantOutputs = variants.map((variant: Record<string, any>) => {
    const document = JSON.parse(JSON.stringify(resolvedModel.document));
    document.module = document.module || {};
    document.module.implementationPolicies = variant.implementationPolicies;
    const variantModel = new SemanticModel(document);
    const lowered = lower(variantModel);
    if (!lowered.artifact) {
      throw new Error(`variant ${variant.id} lowering must include artifact`);
    }
    return {
      id: variant.id,
      derivedArtifact: lowered.artifact,
      canonical: variantModel.canonical(),
      behavior: behaviorResults(variantModel, casesFixture),
    };
  });

  const firstBehavior = JSON.stringify(variantOutputs[0].behavior);
  if (!firstBehavior) {
    throw new Error("cases fixture must include behavior cases");
  }
  if (!variantOutputs.every((item) => JSON.stringify(item.behavior) === firstBehavior)) {
    throw new Error("observable behavior should remain equal across policy variants");
  }
  if (new Set(variantOutputs.map((item) => JSON.stringify(item.derivedArtifact))).size < 2) {
    throw new Error("derived artifacts should differ across policy variants");
  }
  if (!variantOutputs.every((item) => JSON.stringify(item.canonical) === JSON.stringify(variantOutputs[0].canonical))) {
    throw new Error("canonical meaning should remain stable across policy variants");
  }

  return {
    blockedCompilation,
    resolvedCompilation,
    canonicalMeaningStableAcrossVariants: true,
    variants: variantOutputs.map((item) => ({
      id: item.id,
      derivedArtifact: item.derivedArtifact,
    })),
    behaviorCases: variantOutputs[0].behavior.map((item) => item.id),
  };
}

function isDirectExecution(): boolean {
  const entry = process.argv[1];
  return !!entry && import.meta.url === pathToFileURL(resolve(entry)).href;
}

if (isDirectExecution()) {
  console.log(JSON.stringify(runDerivationProof(), null, 2));
}
